### PR TITLE
Txn message

### PR DESCRIPTION
**Pull request by triump2020**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #4693 #4701 

## What this PR does / why we need it:

### CODE REVIEW

File: proto/api.proto [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+import "github.com/gogo/protobuf/gogoproto/gogo.proto";
+import "timestamp.proto";
+import "plan.proto";
+
+
+option go_package = "github.com/matrixorigin/matrixone/pkg/pb/api";
+option (gogoproto.sizer_all) = false;
+option (gogoproto.protosizer_all) = true;
+
+//types.type
```
Comment (fengttt) [inline proto/api.proto:29]: We have decided to not using typeid in proto.   Just use uint8 and use container/types.

File: proto/api.proto [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        Insert = 0;
+        Delete = 1;
+    }
+    EntryType entry_type = 1;
+    uint64    table_id = 2;
+    uint64    database_id = 3;
+    string    table_name = 4;
+    string    database_name = 5;
+    string    file_name = 6;
+    uint64    block_id = 7;
```
Comment (fengttt) [inline proto/api.proto:335]: Adding blocks of data, is one block per entry?   This is rather wasteful.

**Review by fengttt — COMMENTED** [review]
Minor inline comments.

Overall, according to our discussion earlier,

Message contains a list of entries, each entry contains a batch of records, or, a list of S3 blocks which contains records.    We have batch in the doc, but not the S3 option.

Message may also contains a list of blockinfos, which will contains blockid, (s3 path), and a metadata location, and a delete map location.   The blockinfo,  need to be part of the proto file.

Message may contains a checkpoint, which should be a TS, plus either a list of blockinfos, or a S3 object that contains many blockinfos.

File: proto/api.proto [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    //          database_id = dbId
+    //          table_name =  xxx
+    //          database_name = xxx
+    //          file_name = "" 
+    //          block_id = 0
+    //          bat.attrs = {"rowid", "primary column value",  "column1 value", ...}
+    //      }
+
+
+message Entry {
```
[the remark this answers is not in the document]
Comment (nnsgmsone) [reply proto/api.proto:324]: yes

File: proto/api.proto (hunk 2 above)
Comment (nnsgmsone) [reply proto/api.proto:335]: at the moment it is

**Review by ouyuanning — APPROVED** [review]

**Review by yingfeng — APPROVED** [review]

File: proto/api.proto [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+message SyncLogTailReq {
+    //timestamp.Timestamp checkpoint_ts = 1;
+    timestamp.Timestamp cn_have = 1;
+    timestamp.Timestamp cn_want = 2;
+    TableID   table = 3;
+};
+
+message SyncLogTailResp {
+    // location of catalog or metadata checkpoint.
+    string     ckp_location = 1;
```
Comment (nnsgmsone) [inline proto/api.proto:58]: What does this string represent?

File: proto/api.proto [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    Batch     bat = 8;
+};
+
+// Checkpint is a snapshot at a timestamp for catalog and block info, 
+// it contains CatalogCkp and MetadataCkp.
+
+// CatalogCkp contains information about database and tables in the system,and
+// MetadataCkp contains information about blocks.
+message Checkpoint {
+    timestamp.Timestamp min_ts = 1;
```
Comment (nnsgmsone) [inline proto/api.proto:268]: The meaning of max_ts and min_ts is best clarified

File: proto/api.proto [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+    timestamp.Timestamp max_ts = 2;
+    Batch      bat   = 3;
+
+};
+// catalog checkpoint:
+// one Batch represents a table, such as : mo_databases, mo_tables, mo_columns,... etc.
```
Comment (fengttt) [inline proto/api.proto:274]: Write this out in proto, not comment.   Let proto generate code, at least we know it is syntax wise correct.

File: proto/api.proto [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+// one row of Batch represents a block metadata.
+
+// block meta data :
+// type BlockMeta struct {
+//    ID                    uint64
+//    SegmentID             uint64
+//    TableID               uint64
+//    DbID                  uint64
+//    CreatedAt, DeletedAt  timestamp.Timestamp
+//    MetaLoc, DeltaLoc     string
```
Comment (fengttt) [inline proto/api.proto:293]: TAE team will provide method, that given a metaloc, that we can read it from S3.
The content that read out from S3, can be deserialized to go struct -- in fact, should be vector.

File: proto/api.proto [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+        Insert = 0;
+        Delete = 1;
+    }
+    EntryType entry_type = 1;
+    uint64    table_id = 2;
+    uint64    database_id = 3;
+    string    table_name = 4;
+    string    database_name = 5;
+    string    file_name = 6;
+    uint64    block_id = 7;
```
Comment (fengttt) [inline proto/api.proto:258]: Entry maybe Batch of tuples, or, a block.   The block should contain a field  that has the same format as a BlockMeta used below.

File: proto/api.proto (hunk 7 above)
Comment (fengttt) [reply proto/api.proto:293]: Actually, why don't you spec out what people will read out of metaLoc and deltaLoc as well?    A pointer to a go struct is fine.

**Review by XuPeng-SH — APPROVED** [review]
lgtm

**Review by nnsgmsone — APPROVED** [review]

**Review by reusee — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Pull request merged**